File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/notification/ui/NotificationActivity.java
package org.mifos.mobilewallet.mifospay.notification.ui;

import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import org.mifos.mobilewallet.core.domain.model.NotificationPayload;
import org.mifos.mobilewallet.mifospay.R;
import org.mifos.mobilewallet.mifospay.base.BaseActivity;
import org.mifos.mobilewallet.mifospay.notification.NotificationContract;
import org.mifos.mobilewallet.mifospay.notification.presenter.NotificationPresenter;
import org.mifos.mobilewallet.mifospay.utils.Constants;
import org.mifos.mobilewallet.mifospay.utils.DebugUtil;
import org.mifos.mobilewallet.mifospay.utils.Toaster;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * This activity is to view notifications record.
 * Notifications Datatable is populated automatically by server when an event happens.
 * This feature is yet to be implemented on the server side.
 */

public class NotificationActivity extends BaseActivity implements
        NotificationContract.NotificationView {

    @Inject
    NotificationPresenter mPresenter;
    NotificationContract.NotificationPresenter mNotificationPresenter;

    @BindView(R.id.rv_notification)
    RecyclerView mRvNotification;
    @BindView(R.id.tv_placeholder)
    TextView tvplaceholder;

    @Inject
    NotificationAdapter mNotificationAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_notification);
        ButterKnife.bind(this);
        setToolbarTitle("Notifications");
        showColoredBackButton(Constants.BLACK_BACK_BUTTON);
        getActivityComponent().inject(this);

        setupRecyclerView();
        setupSwipeRefreshLayout();

        mPresenter.attachView(this);

        showSwipeProgress();
        mNotificationPresenter.fetchNotifications();
    }

    private void setupRecyclerView() {
        mRvNotification.setLayoutManager(new LinearLayoutManager(this));
        mRvNotification.setAdapter(mNotificationAdapter);
        mRvNotification.addItemDecoration(new DividerItemDecoration(this,
                DividerItemDecoration.VERTICAL));
    }

    private void setupSwipeRefreshLayout() {
        setSwipeRefreshEnabled(true);
        swipeLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                mNotificationPresenter.fetchNotifications();
            }
        });
    }

    @Override
    public void setPresenter(NotificationContract.NotificationPresenter presenter) {
        mNotificationPresenter = presenter;
    }

    @Override
    public void fetchNotificationsSuccess(List<NotificationPayload> notificationPayloadList) {
        hideSwipeProgress();
        if (notificationPayloadList == null || notificationPayloadList.size() == 0) {
            DebugUtil.log("null");
            mRvNotification.setVisibility(View.GONE);
            tvplaceholder.setVisibility(View.VISIBLE);
        } else {
            DebugUtil.log("yes");
            mRvNotification.setVisibility(View.VISIBLE);
            tvplaceholder.setVisibility(View.GONE);
            mNotificationAdapter.setNotificationPayloadList(notificationPayloadList);
        }
        mNotificationAdapter.setNotificationPayloadList(notificationPayloadList);
    }

    @Override
    public void fetchNotificationsError(String message) {
        hideSwipeProgress();
        showToast(message);
    }

    public void showToast(String message) {
        Toaster.showToast(this, message);
    }
}

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/notification/ui/NotificationActivity.kt
package org.mifos.mobilewallet.mifospay.notification.ui

import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import dagger.hilt.android.AndroidEntryPoint
import org.mifos.mobilewallet.core.domain.model.NotificationPayload
import org.mifos.mobilewallet.mifospay.R
import org.mifos.mobilewallet.mifospay.base.BaseActivity
import org.mifos.mobilewallet.mifospay.notification.NotificationContract
import org.mifos.mobilewallet.mifospay.notification.NotificationContract.NotificationView
import org.mifos.mobilewallet.mifospay.notification.presenter.NotificationPresenter
import org.mifos.mobilewallet.mifospay.utils.Constants
import org.mifos.mobilewallet.mifospay.utils.DebugUtil
import org.mifos.mobilewallet.mifospay.utils.Toaster
import javax.inject.Inject

/**
 * This activity is to view notifications record.
 * Notifications Datatable is populated automatically by server when an event happens.
 * This feature is yet to be implemented on the server side.
 */
@AndroidEntryPoint
class NotificationActivity : BaseActivity(), NotificationView {
    @JvmField
    @Inject
    var mPresenter: NotificationPresenter? = null
    var mNotificationPresenter: NotificationContract.NotificationPresenter? = null

    @JvmField
    @BindView(R.id.rv_notification)
    var mRvNotification: RecyclerView? = null

    @JvmField
    @BindView(R.id.tv_placeholder)
    var tvplaceholder: TextView? = null

    @JvmField
    @Inject
    var mNotificationAdapter: NotificationAdapter? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_notification)
        ButterKnife.bind(this)
        setToolbarTitle("Notifications")
        showColoredBackButton(Constants.BLACK_BACK_BUTTON)
        setupRecyclerView()
        setupSwipeRefreshLayout()
        mPresenter?.attachView(this)
        showSwipeProgress()
        mNotificationPresenter?.fetchNotifications()
    }

    private fun setupRecyclerView() {
        mRvNotification?.layoutManager = LinearLayoutManager(this)
        mRvNotification?.adapter = mNotificationAdapter
        mRvNotification?.addItemDecoration(
            DividerItemDecoration(
                this,
                DividerItemDecoration.VERTICAL
            )
        )
    }

    private fun setupSwipeRefreshLayout() {
        setSwipeRefreshEnabled(true)
        swipeRefreshLayout?.setOnRefreshListener { mNotificationPresenter?.fetchNotifications() }
    }

    override fun fetchNotificationsSuccess(notificationPayloadList: List<NotificationPayload?>?) {
        hideSwipeProgress()
        if (notificationPayloadList.isNullOrEmpty()) {
            DebugUtil.log("null")
            mRvNotification?.visibility = View.GONE
            tvplaceholder?.visibility = View.VISIBLE
        } else {
            DebugUtil.log("yes")
            mRvNotification?.visibility = View.VISIBLE
            tvplaceholder?.visibility = View.GONE
            mNotificationAdapter?.setNotificationPayloadList(notificationPayloadList as List<NotificationPayload>)
        }
        mNotificationAdapter?.setNotificationPayloadList(notificationPayloadList as List<NotificationPayload>)
    }

    override fun fetchNotificationsError(message: String?) {
        hideSwipeProgress()
        showToast(message)
    }

    override fun setPresenter(presenter: NotificationContract.NotificationPresenter?) {
        mNotificationPresenter = presenter
    }

    fun showToast(message: String?) {
        Toaster.showToast(this, message)
    }
}